cpp/include/cudf/join.hpp

```diff
@@ -213,6 +213,85 @@ std::unique_ptr<cudf::experimental::table> full_join(
                          rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief  Performs an inner join using the nested loop join implementation.
+ *
+ * Nested loop join computes the cross product of the two tables (compares
+ * all rows in Left to all rows in Right) and filters the product based upon
+ * a provided filtering function which identifies the rows that will be part
+ * of the output join.
+ *
+ * @example Left a: {0, 1, 2}
+ *          Right b: {1, 2, 3}, a: {1, 2, 5}
+ *          Filter compares Left[0] to Right[1], returns true if they match
+ *          columns_in_common: { {0, 1} }
+ * Result: { a: {1, 2}, b: {1, 2} }
+ *
+ * @example Left a: {0, 1, 2}
+ *          Right b: {1, 2, 3}, c: {1, 2, 5}
+ *          Filter compares Left[0] to Rignt[0], returns true if they match
+ *          columns_in_common: { }
+ * Result: { a: {1, 2}, b: {1, 2}, c: {1, 2} }
+ *
+ * @throws cudf::logic_error if `columns_in_common` contains a pair of indices
+ * (L, R) if L does not exist in `left_on` or R does not exist in `right_on`.
+ * @throws cudf::logic_error if `columns_in_common` contains a pair of indices
+ * (L, R) such that the type of `L` within `left_on` is not equal to
+ * type of R within `right_on`
+ * @throws cudf::logic_error if number of elements in `left_on` or `right_on`
+ * mismatch.
+ * @throws std::out_of_range if element of `left_on` or `right_on` exceed the
+ * number of columns in the left or right table.
+ *
+ * @param[in] left               The left table
+ * @param[in] right              The right table
+ * @param[in] columns_in_common  A vector of pairs of column indices into
+ * `left` and `right`, respectively, that are "in common". For "common"
+ * columns, only a single output column will be produced, which is gathered
+ * from the `left` table.
+ * @param[in] filter             A function that accepts a left index and
+ *                               right index corresponding to rows in the
+ *                               left and right tables.  Returns true if
+ *                               the row is part of the resulting join, false
+ *                               if not
+ * @param mr                     Memory resource used to allocate the returned
+ *                               table and columns
+ *
+ * @returns    Result of joining `left` and `right` tables, filtered by the
+ *             provided filter. The resulting table will be joined columns of
+ *             `left(including common columns)+right(excluding common columns)`.
+ */
+template <typename Filter>
+std::unique_ptr<table> nested_loop_inner_join(cudf::table_view const& left,
+                                              cudf::table_view const& right,
+                                              std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                                              Filter filter,
```

Comment:
Would it be possible to provide a default parameter for the filter? From what I know, equality is commonly used here so it should benefit many users to default to this.

Reply:
I agree. There should be an 'identity' filter that should not do anything.

Reply:
Yes, I imagine an identity filter that always returns true.  For equality (where this is starting), I imagine a predefined filter that identifies a list of column pairs and does the rest of the join computation (return true where the columns in left match the columns in right).

When we get to inequality joins, the filter could be much more varied and I expect we might define a few others.

Of course, ultimately the filter could be used for non-join filtering as well.  But those would all be custom built.

Reply:
Could you add some example of a Filter function and what it could look like?

Reply:
Yes.  I'll try and push something with an example.

Reply:
Abandoned the idea of using the filter approach.  Everything would need to be templated all they way up to the interface.  That seems like too much complexity for little benefit.